api/v1/server/handlers/workflows/cancel.go

```diff
@@ -0,0 +1,103 @@
+	"github.com/hatchet-dev/hatchet/internal/repository/prisma/dbsqlc"
+	"github.com/hatchet-dev/hatchet/internal/services/shared/tasktypes"
+)
+
+func (t *WorkflowService) WorkflowRunCancel(ctx echo.Context, request gen.WorkflowRunCancelRequestObject) (gen.WorkflowRunCancelResponseObject, error) {
+	tenant := ctx.Get("tenant").(*db.TenantModel)
+	runIds := request.Body.WorkflowRunIds
+
+	canceledStepRunsChannel := make(chan []*db.StepRunModel, len(runIds))
+	var returnErr error
+
+	for _, runId := range runIds {
+		go func(runId uuid.UUID) {
+			// Lookup step runs for the workflow run
+			runIdStr := runId.String()
+			stepRuns, err := t.config.EngineRepository.StepRun().ListStepRunsByWorkflowRunId(ctx.Request().Context(), tenant.ID, runIdStr)
+			if err != nil {
+				returnErr = multierror.Append(err, fmt.Errorf("failed to list step runs for workflow run %s", runIdStr))
+				canceledStepRunsChannel <- nil
+				return
+			}
+
+			var canceledStepRuns []*db.StepRunModel
+
+			for _, stepRun := range stepRuns {
+				status := stepRun.StepRun.Status
+
+				// If the step run is not in a final state, send a task to the taskqueue to cancel it
+				if status != dbsqlc.StepRunStatusSUCCEEDED && status != dbsqlc.StepRunStatusFAILED && status != dbsqlc.StepRunStatusCANCELLED {
+					var reason = "CANCELLED_BY_USER"
+					// send a task to the taskqueue
+					err = t.config.MessageQueue.AddMessage(
+						ctx.Request().Context(),
+						msgqueue.JOB_PROCESSING_QUEUE,
+						tasktypes.StepRunCancelToTask(stepRun, reason),
+					)
+					if err != nil {
+						returnErr = multierror.Append(err, fmt.Errorf("failed to send cancel task for step run %s", pgUUIDToStr(stepRun.StepRun.ID)))
+						continue
+					}
+
+					stepRunId := pgUUIDToStr(stepRun.StepRun.ID)
+					stepRunDb, err := t.config.APIRepository.StepRun().GetStepRunById(tenant.ID, stepRunId)
+					if err != nil {
+						returnErr = multierror.Append(err, fmt.Errorf("failed to get step run %s", stepRunId))
+						continue
+					}
+
+					// Add the canceled step run to the list
+					canceledStepRuns = append(canceledStepRuns, stepRunDb)
+				}
+			}
+
+			canceledStepRunsChannel <- canceledStepRuns
+		}(runId)
+	}
+
+	var allCanceledStepRuns []*db.StepRunModel
+	for i := 0; i < len(runIds); i++ {
+		canceledStepRuns := <-canceledStepRunsChannel
+		if canceledStepRuns != nil {
+			allCanceledStepRuns = append(allCanceledStepRuns, canceledStepRuns...)
+		}
+	}
+
+	if returnErr != nil {
+		return nil, returnErr
+	}
+
+	// Transform the canceled step runs to the response format
+	canceledStepRunsResponse := make([]gen.StepRun, len(allCanceledStepRuns))
+	for i, stepRun := range allCanceledStepRuns {
+
+		res, err := transformers.ToStepRun(stepRun)
+
+		if err != nil {
+			return nil, err
+		}
+
+		canceledStepRunsResponse[i] = *res
+	}
+
+	// Return the list of canceled step runs in the response
+	return gen.WorkflowRunCancel200JSONResponse(canceledStepRunsResponse), nil
```

Comment:
The API design here isn't very intuitive. It doesn't make sense to me that the resource which is cancelled is a workflow run, but a list of step runs are returned from this endpoint. It might be better to use `ListWorkflowRuns` and pass the array of workflow run IDs, and just return the list of workflow runs which were cancelled.